### PR TITLE
Feature/plmc 425 projectmetadata revisited

### PR DESCRIPTION
**Pull request by JuaniRios**
## What?
1. ProjectMetadata changes:
	1. `ticket_size` renamed to `round_ticket_sizes` . It means something different now. We have 2 auction ticket sizes for inst. and prof. credentials, and 3 contribution ticket sizes for each credential. The minimum refers to the min CT to buy per participation, the max refers to the max CT a single DID can buy. Bidders cannot bid below 5k USD
	
	2. `total_allocation_size` now refers again to the total size and is not split into 2, one for auction and one for contribution
	
	3. new field called `auction_round_allocation_percentage`. This tells us how many CTs are available in the auction.

	4. `participation_currencies` now accepts a vector of currencies
	
	5. `participants_size`, `funding_thresholds` removed till we have time to implement them post-mvp. `conversion_rate` removed since it was no longer used

2. ProjectDetails changed: `remaining_contribution_tokens` is now a single field
3. DID added to `UntrustedToken`
4. e2e tests broken
5. do_functions now require DID and investor type
6. All Investor type extractors from the Untrusted token are grouped into a single config type
7. Default values in tests changed

## Why?
1. We now have credentials and investor types. We had a lot of metadata fields waiting for those to be implemented.
	1. The ticket sizes that we asked potential investors were different than the ones we thought we were using, so we had to adapt the logic
	2. We wanted to be able to use the unsold bid CTs in the contribution rounds
	3. Needed for point 2.
	4. Long overdue
	5. We don't have time to implement those, and they are required for RWAs and not CTs
2. `total_allocation_size` is also a single field, so we don't need 2 fields here.
3. Max ticket sizes are per did and not per account
4. The new min ticket of 5k USD was not considered when writing that.
5. We need them to store the DID to participation count mapping, and the TicketSize check
6. Made code cleaner
7. Previous ones had bids below 5kUSD

## How?
1. ProjectMetadata
	1. New enums with generic fields to implement a constraint at the time of project creation.
	   For example, all ticket sizes in the auction round cannot be created with a min CT amount below 5k USD. (note we need the min_price for this calculation). 
	2. It is now just one `Balance` field
	3. It is a `Percentage`, and you use it with the `total_allocation_size` to get the amount available for the auction round
	4. We put a `BoundedVec` of `AcceptedFundingAsset` with a length of the total amount of variants. We also have a validity check that we have no duplicates
	5. Just remove the fields from the struct
2. Change it to `Balance` and fix all calculations.
3. Add the field to the `UntrustedToken` struct, and pass it down to the inner extrinsic along with the investor type. The DID is now stored on chain along with the number of participations.
4. We add an ignore flag on those tests till we have new e2e values to test
5. We use the values from the jwt on extrinsics, and pass mocked values on tests
6. The function accepts a jwt, and returns the investor type and did if the token is valid. 
7. A lot of changes to testing logic had to be made, could be skipped maybe if reviewers don't have enough time to go through the 5k line changes :c

## Testing?
New tests:
- setting bid ticket below 5k usd in terms of CT should fail
- tokens not bought as part of the auction allocation are available in the community and remainder round
- ticket sizes are respected in all rounds
- issuers can declare multiple funding currencies, and participants can use all of them.

All pallet tests including benchmarks + integration tests should be run since these changes affected them all.

## Anything Else?
- We will revisit the project metadata in the future to add the thresholds and participant sizes.
- E2E tests are broken, and will be fixed once we have the new test values.

### CODE REVIEW

File: pallets/funding/src/tests.rs
Comment (vstam1) [inline pallets/funding/src/tests.rs:6328]: remove?

File: rust-toolchain.toml
Comment (vstam1) [inline rust-toolchain.toml:2]: Should probably not change toolchain all the time?

**Review by vstam1 — CHANGES_REQUESTED** [review]
Some small and some larger changes. Open for discussion about certain points. I also want to take a second look at the Ticket size implementation. It feels overly complicated with all the trait bounds etc while the logic should be relatively simple. Might have overlooked something.

File: integration-tests/Cargo.toml
Comment (JuaniRios) [reply integration-tests/Cargo.toml:137]: Used zepter to try and fix things. We can remove it

File: polimec-common/common/src/credentials/mod.rs
Comment (JuaniRios) [reply polimec-common/common/src/credentials/mod.rs:126]: I prefer it since it makes it easier to debug. What do you think @lrazovic ?

File: pallets/funding/src/functions.rs
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:80]: I expected there to be more features to include that need credentials, so just in case passing them to use later on. 
If we already agree credential use to be as it is, we can remove them yes

File: pallets/funding/src/functions.rs
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:198]: If an issuer cannot evaluate their own project, then it makes sense to use their DID no?
Otherwise why look at their account if they can easily create new ones

File: pallets/funding/src/functions.rs
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:907]: Well we should probably check the credential as well as the issuer account no?
In case their credential expired

File: pallets/funding/src/functions.rs
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:936]: True, and the extrinsic evaluate already checks they have a valid credential.

File: pallets/funding/src/functions.rs
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:1132]: I like it

File: pallets/funding/src/functions.rs
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:1428]: 👍

File: pallets/funding/src/functions.rs
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:2246]: I tried to be consistent in that if some extrinsics needed to pass the did and credential to the inner function, then do the investor type check also in the inner function. And then if some extrinsics behave like that, then make all extrinsics pass the did + investor type and do the checks there.

if you think its more readable to have some extrinsics do investor type checks on the outside and some on the inside then I'll change it

File: pallets/funding/src/benchmarking.rs
Comment (JuaniRios) [reply pallets/funding/src/benchmarking.rs:2771]: nope

File: pallets/funding/src/instantiator.rs
Comment (JuaniRios) [reply pallets/funding/src/instantiator.rs:836]: Sure

File: rust-toolchain.toml
Comment (JuaniRios) [reply rust-toolchain.toml:2]: Should have it in my .gitignore, sry

File: polimec-common/common/src/credentials/mod.rs
Comment (vstam1) [reply polimec-common/common/src/credentials/mod.rs:126]: You mean for placing dbg between let result and result return? I think it would be proper to actually directly return the result (probably clippy will also start complaining), and only do a quick rewrite if you want to debug the result and move it back once debug is not necessary anymore

File: pallets/funding/src/functions.rs
Comment (vstam1) [reply pallets/funding/src/functions.rs:80]: I think its mostly as is now. We can always pass them later if we actually need them. Its inner functions so nothing for extrinsics will change, so its a simpel runtime upgrade if we want to add additional logic. Lets keep is simpel for now

File: pallets/funding/src/functions.rs
Comment (vstam1) [reply pallets/funding/src/functions.rs:198]: Yeah might make sense. Maybe something to discuss with Luca? I think it would be fine to store the Did in project_metadata/details and check for it in for example evaluation. At the same time, I don't think we need it here, right? As it actually only starts evaluation and does not do an evaluation?

File: pallets/funding/src/functions.rs
Comment (vstam1) [reply pallets/funding/src/functions.rs:907]: Can probably also do in outer call? Same for all the other "issuer" related functions

File: pallets/funding/src/functions.rs
Comment (vstam1) [reply pallets/funding/src/functions.rs:2246]: I understand your reasoning, but I am always in favor of simplifying parameter passing. I also think we could potentially do the investor type checks in every outer function, and pass the investor type only to the functions that actually need it for other logic. That way we keep all the investor type checks in the outer functions.

File: polimec-common/common/src/credentials/mod.rs
Comment (JuaniRios) [reply polimec-common/common/src/credentials/mod.rs:126]: I mean for using the debugger, but sure

File: pallets/funding/src/functions.rs
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:198]: agreed, I created a Linear issue for this, but as mentioned this place does not need the did

File: pallets/funding/src/functions.rs
Comment (vstam1) [inline pallets/funding/src/functions.rs:1326]: Yeah love these changes

**Review by vstam1 — APPROVED** [review]
Would still like to see the ticket sizes simplified, but let that up to @lrazovic and won't block the pr for it. Rest looks good to me!

File: pallets/funding/src/lib.rs
Comment (lrazovic) [inline pallets/funding/src/lib.rs:1033]: I'm not entirely convinced about the new `ensure_origin` syntax, especially with the repetitive use of ensure (`ensure_origin` + `ensure!`). I get that in the Evaluation extrinsics, where we need to check whether the origin is Institutional OR Professional OR Retail, the old syntax feels a bit clunky. However, I believe we might strike a balance. What if we incorporate the `investor_type` check within the `ensure_origin` function and simply pass the minimum required investor type to it? We could introduce a hypothetical investor type called `Any` for instances where we're open to any credentialed user. Alternatively, we could pass an array of `InvestorTypes` or even use numerical representation for Retail (0), Professional (1), and Institutional (2) to denote the minimum required "rank". In summary: I think there are several ways we could explore to refine the current syntax even further.

File: pallets/funding/src/types.rs
Comment (lrazovic) [inline pallets/funding/src/types.rs:183]: We can now remove/rephrase the `(Auction, Community)` tuple from the comment.

File: pallets/funding/src/types.rs
Comment (lrazovic) [inline pallets/funding/src/types.rs:660]: We can now remove this

File: pallets/funding/src/types.rs
Comment (lrazovic) [inline pallets/funding/src/types.rs:986]: Is there no other possible way to avoid having a new custom type?

File: pallets/funding/src/types.rs
Comment (lrazovic) [inline pallets/funding/src/types.rs:696]: Do we still need this even if we use the `VariantCount`?

**Review by lrazovic — CHANGES_REQUESTED** [review]
Like all massive PR, it is not easy to do a review, and I definitely missed something. 

I'm pleased to see the `ticket_size` is now a single value again, with a portion of it allocated for the Auction Round. It seems to fit better with the front-end too. However, I find the new boundary controls quite complex. It would be great to have them in a simpler yet equally effective format, as validating user input is crucial. Sadly, I don't have a simpler solution to suggest right now.

I believe there's room for improvement with the new `ensure_origin`, and I've shared more thoughts in the comment.

On a general note, I'd love to see more [`DocComments`](https://doc.rust-lang.org/rust-by-example/meta/doc.html#doc-comments) in the functions, particularly in the various `is_valid`s. It would really help us understand the purpose of each `is_valid` for a given struct, aiding not just our current understanding but also our future selves.

There are also a few unused imports, as revealed by running cargo c -r -p pallet-funding, which should be addressed.

File: pallets/funding/src/types.rs
Comment (lrazovic) [reply pallets/funding/src/types.rs:183]: I'm actually referring to

`/// Total allocation of Contribution Tokens available for the Funding Round. (Auction, Community)`

File: pallets/funding/src/types.rs
Comment (lrazovic) [reply pallets/funding/src/types.rs:660]: I'm actually referring to

`// TODO: PLMC-157. Use SCALE fixed indexes`

File: polimec-common/common/src/credentials/mod.rs
Comment (JuaniRios) [reply polimec-common/common/src/credentials/mod.rs:167]: Done in an upstream PR

File: polimec-common/common/src/credentials/mod.rs
Comment (vstam1) [inline polimec-common/common/src/credentials/mod.rs:65]: One small this. I think the naming convention for types should be PascalCase so might be good to change it to Did :)

Comment (vstam1) [reply polimec-common/common/src/credentials/mod.rs:65]: But at the same time its an abbreviation so might be actually correct casing haha. Okay skip my previous remark 🤡

Comment (JuaniRios) [reply polimec-common/common/src/credentials/mod.rs]: CleanShot 2024-03-15 at 11 44 47@2x

ChatGPT agreed with you tho, which is why I changed it lol

File: polimec-common/common/src/credentials/mod.rs
Comment (lrazovic) [reply polimec-common/common/src/credentials/mod.rs:186]: It should be already done in one of the many stacked PR https://github.com/Polimec/polimec-node/pull/177#discussion_r1524600465

Comment (JuaniRios) [reply polimec-common/common/src/credentials/mod.rs:186]: It was, but introduced it down too. Tried in test-utils and gave me a headache with the feature flags so put it in common.
I'm moving it to test-utils now 😤

**Review by lrazovic — APPROVED** [review]
After several review rounds: LGTM

**Pull request merged**
